torchdata/dataloader2/communication/iter.py

```diff
@@ -149,53 +160,117 @@ def DataPipeBehindQueues(
+                seed_generator = request.seed_generator
+                iter_reset_fn = request.iter_reset_fn
+                dispatching_dps = find_dps(traverse_dps(source_datapipe), _IterateQueueDataPipes)
+                for dp in dispatching_dps:
+                    dp.reset_epoch(seed_generator, iter_reset_fn)
+                source_datapipe = process_reset_fn(
+                    source_datapipe,
+                    worker_info,
+                    seed_generator,
+                    distributed_shared_seed,
+                    iter_reset_fn,
+                    custom_reset_fn,
+                )
+            source_datapipe.reset_iterator()
+            if request_counter is not None:
+                request_counter.reset("reset_epoch")
+                while request_counter.is_reached("reset_epoch"):
+                    yield True
             protocol.response_reset_epoch()
+            yield True  # Returns control
 
-        elif isinstance(request, communication.messages.ResetIteratorRequest):
-            # Ensure only reset iterator once for the dispatching process
-            if reset_iterator_counter is not None:
-                reset_iterator_counter.increment()
-                while not reset_iterator_counter.is_reached():
+        elif isinstance(request, communication.messages.LimitRequest):
+            if request_counter is not None:
+                request_counter.increment("limit")
+                # Make sure all loops have reached
+                while not request_counter.is_reached("limit"):
                     yield True
-                # Sync between loops within the dispatching process
-                source_datapipe.reset_iterator()
-                yield True
-                reset_iterator_counter.reset()
-            source_datapipe.reset_iterator()
-            protocol.response_reset_iterator()
+                # yield True
+            if request_counter is None or loop_id == 0:
+                num_batches = request.num_batches
+                limit_fn = request.limit_fn
+                worker_num_batches = num_batches if request.worker_num_batches is None else request.worker_num_batches
+                # Send limit to the worker/dispatching process
+                dispatching_dps = find_dps(traverse_dps(source_datapipe), _IterateQueueDataPipes)
+                for dp in dispatching_dps:
+                    dp.request_limit(num_batches, limit_fn, worker_num_batches)
+                if limit_fn is not None:
+                    # Set limit to the DataPipe graph in worker/dispatching process
+                    source_datapipe = limit_fn(source_datapipe, worker_num_batches)
+            if request_counter is not None:
+                request_counter.reset("limit")
+                while request_counter.is_reached("limit"):
+                    yield True
+            protocol.response_limit()
+            yield True  # Returns control
```

Comment:
nit: I wonder if we can have a helper function to handles all the overlapping logic among `limit/pause/resume` with more inline docs.

Reply:
Yeah, I can do that.